### PR TITLE
[Merged by Bors] - feat(Algebra/Homology): signs for the associativity compatibility of the total complex

### PR DESCRIPTION
**Pull request by joelriou**
This PR introduces a typeclass `ComplexShape.Associative` which takes as inputs six complex shapes `c₁`, `c₂`, `c₃`, `c₁₂`, `c₂₃` and `c`, and asserts compatibilities of signs for the associativity of the total complex of a triple complex. It shall also be used in the construction of the associator isomorphism for the tensor product of homological complexes.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Homology/ComplexShapeSigns.lean [hunk 1]
```diff
@@ -183,6 +183,85 @@ lemma ε_up_ℤ (n : ℤ) : (ComplexShape.up ℤ).ε n = n.negOnePow := rfl
...
+variable [TotalComplexShape c₁₂ c₃ c] [TotalComplexShape c₂ c₃ c₂₃] [TotalComplexShape c₁ c₂₃ c]
+
+/-- When we have three complexes shapes `c₁`, `c₂`, `c₃`, `c₁₂`, `c₂₃`, `c`, and total functors
+`HomologicalComplex₂ C c₁ c₂ ⥤ HomologicalComplex C c₁₂`,
+`HomologicalComplex₂ C c₁₂ c₃ ⥤ HomologicalComplex C c`,
+`HomologicalComplex₂ C c₂ c₃ ⥤ HomologicalComplex C c₂₃`,
+`HomologicalComplex₂ C c₁ c₂₂₃ ⥤ HomologicalComplex C c`, we get two ways to
+compute the total complex of a triple complex in `HomologicalComplex₃ C c₁ c₂ c₃`, then
+under this assumption `[Associator c₁ c₂ c₃ c₁₂ c₂₃ c]`, these two complexes
+canonical identify (without introducing signs). -/
```
Comment (dagurtomas) [inline Mathlib/Algebra/Homology/ComplexShapeSigns.lean:198]:
```suggestion
/-- When we have sic complex shapes `c₁`, `c₂`, `c₃`, `c₁₂`, `c₂₃`, `c`, and total functors
`HomologicalComplex₂ C c₁ c₂ ⥤ HomologicalComplex C c₁₂`,
`HomologicalComplex₂ C c₁₂ c₃ ⥤ HomologicalComplex C c`,
`HomologicalComplex₂ C c₂ c₃ ⥤ HomologicalComplex C c₂₃`,
`HomologicalComplex₂ C c₁ c₂₂₃ ⥤ HomologicalComplex C c`, we get two ways to
compute the total complex of a triple complex in `HomologicalComplex₃ C c₁ c₂ c₃`, then
under this assumption `[Associator c₁ c₂ c₃ c₁₂ c₂₃ c]`, these two complexes
canonically identify (without introducing signs). -/
```

File: Mathlib/Algebra/Homology/ComplexShapeSigns.lean [hunk 2]
```diff
@@ -183,6 +183,85 @@ lemma ε_up_ℤ (n : ℤ) : (ComplexShape.up ℤ).ε n = n.negOnePow := rfl
...
+class Associator : Prop where
```
Comment (dagurtomas) [inline Mathlib/Algebra/Homology/ComplexShapeSigns.lean:199]: The name `Associator` sounds like the name of an isomorphism, but this is a `Prop`, so maybe `Associative` is a better name?

**Review by dagurtomas — APPROVED** [review]

File: Mathlib/Algebra/Homology/ComplexShapeSigns.lean (hunk 2 above)
Comment (joelriou) [reply Mathlib/Algebra/Homology/ComplexShapeSigns.lean]: Thanks for the suggestion! I have changed it to `Associative`.

**Comment by dagurtomas** [comment]
Thanks!

maintainer merge

**Review by TwoFX — APPROVED** [review]
You still have the old name `Associator` in the PR description/future commit message.

Thanks!
bors r+

**Comment by TwoFX** [comment]
bors d+

**Comment by joelriou** [comment]
Thanks! Retrying...

bors merge

**Pull request closed without merging**